packages/xstate-store/src/store.ts

```diff
@@ -0,0 +1,273 @@
+  let currentSnapshot = initialSnapshot;
+  const transition = createStoreTransition<TContext, TEventPayloadMap>(
+    transitions,
+    updater
+  );
+
+  function receive(event: StoreEvent) {
+    currentSnapshot = transition(currentSnapshot, event);
+
+    observers?.forEach((o) => o.next?.(currentSnapshot));
+  }
+
+  const store: Store<TContext, StoreEvent> = {
+    send(event) {
+      receive(event as unknown as StoreEvent);
+    },
+    getSnapshot() {
+      return currentSnapshot;
+    },
+    getInitialSnapshot() {
+      return initialSnapshot;
+    },
+    subscribe(observerOrFn) {
+      const observer = toObserver(observerOrFn);
+      observers ??= new Set();
+      observers.add(observer);
+
+      return {
+        unsubscribe() {
+          return observers?.delete(observer);
+        }
+      };
+    },
+    [symbolObservable](): InteropSubscribable<StoreSnapshot<TContext>> {
+      return this;
+    }
+  };
+
+  return store;
+}
+
+/**
+ * Creates a `Store` that has its own internal state and can receive events.
+ *
+ * @example
+  ```ts
+  const store = createStore({
+    // initial context
+    { count: 0 },
+    // transitions
+    {
+      on: {
+        inc: (context, event: { by: number }) => {
+          return {
+            count: context.count + event.by
+          }
+        }
+      }
+    }
+  });
+
+  store.getSnapshot(); // { context: { count: 0 } }
+  store.send({ type: 'inc', by: 5 });
+  store.getSnapshot(); // { context: { count: 5 } }
+  ```
+ */
+export function createStore<
+  TContext extends StoreContext,
+  TEventPayloadMap extends EventPayloadMap
+>(
+  initialContext: TContext,
+  transitions: {
+    [K in keyof TEventPayloadMap & string]:
+      | StoreAssigner<NoInfer<TContext>, { type: K } & TEventPayloadMap[K]>
+      | StorePropertyAssigner<
+          NoInfer<TContext>,
+          { type: K } & TEventPayloadMap[K]
+        >;
+  }
+): Store<TContext, ExtractEventsFromPayloadMap<TEventPayloadMap>> {
+  return createStoreCore(initialContext, transitions);
+}
+
+/**
+ * Creates a `Store` with a provided producer (such as Immer's `producer(…)`
+ * A store has its own internal state and can receive events.
+ *
+ * @example
+  ```ts
+  import { produce } from 'immer';
+
+  const store = createStoreWithProducer(produce, {
+    // initial context
+    { count: 0 },
+    // transitions
+    {
+      on: {
+        inc: (context, event: { by: number }) => {
+          context.count += event.by;
+        }
+      }
+    }
+  });
+
+  store.getSnapshot(); // { context: { count: 0 } }
+  store.send({ type: 'inc', by: 5 });
+  store.getSnapshot(); // { context: { count: 5 } }
+  ```
+ */
+export function createStoreWithProducer<
+  TContext extends StoreContext,
+  TEventPayloadMap extends EventPayloadMap
+>(
+  producer: (
+    context: TContext,
+    recipe: (context: TContext) => void
+  ) => TContext,
+  initialContext: TContext,
+  transitions: {
+    [K in keyof TEventPayloadMap & string]: (
```

Comment:
I have a strong feeling that this wont work with TS versions earlier than 5.4. It's a new package though so I guess it's fine.

Reply:
And you are using `NoInfer` so it already requires 5.4 anyway